### PR TITLE
Game Lab: Test initial animations

### PR DESCRIPTION
**Pull request by islemaster**
Adding better animation data validation, unit tests for the animation data validation, and UI tests for levels with initial animations.

### CODE REVIEW

**Comment by islemaster** [comment]
@pcardune so you can be picky with my code this time.

File: apps/src/gamelab/PropTypes.js [hunk 1]
```diff
@@ -132,17 +132,57 @@ export function throwIfSerializedAnimationListIsInvalid(serializedAnimationList)
     throw new Error('Animation List orderedKeys should be an array');
   }
```
Comment (pcardune) [inline apps/src/gamelab/PropTypes.js:172]: the above checks make me want to just use flowtype

Comment (pcardune) [inline apps/src/gamelab/PropTypes.js:172]: If only https://www.npmjs.com/package/react-prop-types-check were just a little bit more active.

File: apps/src/gamelab/PropTypes.js [hunk 2]
```diff
@@ -132,17 +132,57 @@ export function throwIfSerializedAnimationListIsInvalid(serializedAnimationList)
...
+  for (let i = propsNotInOrderedKeys.length - 1; i >= 0; i--) {
+    let j = orderedKeysNotInProps.indexOf(propsNotInOrderedKeys[i]);
+    if (j !== -1) {
+      propsNotInOrderedKeys.splice(i, 1);
+      orderedKeysNotInProps.splice(j, 1);
+    }
+  }
+  if (orderedKeysNotInProps.length > 0) {
+    throw new Error('Animation List has ' +
+        (orderedKeysNotInProps.length === 1 ? 'key' : 'keys') + ' ' +
```
Comment (pcardune) [inline apps/src/gamelab/PropTypes.js:196]: that's dedication to easy to read error messages right there...

File: apps/src/gamelab/PropTypes.js [hunk 3]
```diff
@@ -132,17 +132,57 @@ export function throwIfSerializedAnimationListIsInvalid(serializedAnimationList)
...
+        orderedKeysNotInProps.map(k => `"${k}"`).join(', ') +
+        ' but not associated props');
+  }
+  if (propsNotInOrderedKeys.length > 0) {
+    throw new Error('Animation List has a props for ' +
+        propsNotInOrderedKeys.map(k => `"${k}"`).join(', ') +
+        ' but ' +
+        (propsNotInOrderedKeys.length === 1 ? "that key isn't" : "those keys aren't") +
```
Comment (pcardune) [inline apps/src/gamelab/PropTypes.js:204]: 👏

**Comment by pcardune** [comment]
Have you considered using Immutable.js `OrderedMap` objects? https://facebook.github.io/immutable-js/docs/#/OrderedMap

Seems like that might solve some of these problems for you.

File: apps/src/gamelab/PropTypes.js (hunk 1 above)
Comment (islemaster) [inline apps/src/gamelab/PropTypes.js:172]: Given that [it's all of ten lines of code](https://github.com/dmitriiabramov/react-prop-types-check/blob/master/src/react_prop_types_check.js) I think we could take a similar local approach (much like your tests over your custom prop validators do).  This is a good idea, I'll run with it.

**Comment by islemaster** [comment]
I'm not sure `OrderedMap` is exactly what I want, or maybe I'm misunderstanding it; I want the separate `orderedKeys` array so I can easily reorder animations in the project.  With OrderedMap I'd have to remove all elements after the change point and reinsert them?

**Comment by pcardune** [comment]
@islemaster OrderedMap just gives you an object that behaves both like a Map and a List and supports the APIs of both maps and lists... So you can reorder an OrderedMap just as easily as you can reorder a list...

**Comment by islemaster** [comment]
Okay, so a switch to `OrderedMap` may be useful, along with using [transit-immutable-js](https://github.com/glenjamin/transit-immutable-js) to serialize/deserialize it.  There's likely to be some pain in the transition though, both because we need to migrate any existing levels and projects, and because the levelbuilder interface to this stuff is raw JSON right now and I don't think I want to force levelbuilders to write transit-compatible JSON for initial animations.  Maybe we can make this change as part of getting a nicer levelbuilder animations interface in place?

In any case, having these tests and validation in place seems like a good thing to do before changing the underlying data structure.

File: dashboard/test/ui/step_definitions/steps.rb [hunk 4]
```diff
@@ -454,7 +454,7 @@ def color_for_status(status)
 end
 
 Then /^element "([^"]*)" does not exist/ do |selector|
-  @browser.execute_script("return $(#{selector.dump}).length").should eq 0
+  expect(@browser.execute_script("return $(#{selector.dump}).length")).to eq 0
```
Comment (islemaster) [inline dashboard/test/ui/step_definitions/steps.rb:457]: Note: Switching to rspec's new expect syntax here (the `should` syntax is deprecated, we've been getting warnings about it).

**Comment by pcardune** [comment]
LGTM

**Pull request merged**